README.md
[![Build Status](https://github.com/penguin-teal/clhasher/actions/workflows/Build.yml/badge.svg)](https://github.com/penguin-teal/clhasher/actions/workflows/Build.yml)

# clhasher
CLHasher is a Linux command line utility to get non-cryptographic hashes
of strings or data.

```shell
clhasher <options> <algorithm flag> <input>
```

- [Installing](#installing)
- [Updating](#updating)
- [Uninstalling](#uninstalling)
- [Use](#use)
- [Algorithms](#algorithms)
- [Options](#options)
- [Escaping](#escaping)

## Installing

1. Clone the repository.
```shell
# Clone into directory 'clhasher'
git clone https://github.com/penguin-teal/clhasher.git clhasher
# Enter that directory
cd clhasher
# Go to specific release (v1.1.0 is latest)
git switch --detach v1.1.0
```

2. Build the project.
```shell
# Make a release build
make release
```

3. Install. To install there is a `make install` target and an optional `DESTDIR` variable.
```shell
# Install to /usr/local/bin: (ROOT is required)
sudo make install
# or to other dir:
# make install DESTDIR=/some/other/dir
```

## Updating

```shell
# Go to where you cloned the repository
cd clhasher
# Get the changes
git pull
# Switch to latest (v1.1.0 is latest)
git switch --detach v1.1.0
# Clean just to be safe then build release again
make clean release
# Install like you did before (ROOT is required)
sudo make install
```

## Uninstalling

To uninstall just `rm` the executable file:
```shell
# if you used make install: (ROOT is required)
rm /usr/local/bin/clhasher -i
# or if you installed using DESTDIR=*:
# rm /some/other/dir/clhasher -i
```

## Use

This creates a 64-bit hash using the FNV-1A algorithm:
```shell
clhasher --fnv1a-64 'My String'
```
Which prints:
```
9354639888599838980
```

You can also format the output better:
```shell
clhasher -ax --fnv1a-64 'My String'
```
Which prints:
```
u64 [0]:    0x81D25B6A69B17104
```

The `-a` option is for `--annotate` which produces a little more of a formatted
output, and `-x` is for `--hex` which prints the hash as a base-16 hexadecimal
number.

You can also split large integers into smaller ones:
```shell
clhasher -ax --fnv0-128 --split-64 'splitting'
```

```
u64 [0]:        0x2FE83015B8E8ABEC
u64 [8]:        0x12A4876D7700025E
```

It can also read from STDIN:
```shell
gcc --version | grep --color=never 'gcc' | clhasher -x --fnv1a-64
```

```
DBEF2942FEC1D09F
```

More examples:
```shell
# Hash entire file:
clhasher -x -i ./Makefile --fnv1-32
# 50C5D52

# Hash up to 20 chars from a file:
clhasher -x -i ./Makefile --fnv1a-32 -z 20
# 577B9B35

# Hash the NUL terminator:
clhasher -x0 --fnv1a-64 'TESTSTR'
# BEB7B33DE037BF70

# Use escape codes:
clhasher -xe --fnv1a-64 'line1\n\tline2'
# B2063D2905CD97E1

# Use multiline hashing:
clhasher -axm --fnv1a-64
# > Hello
# #01  u64 [0]:	0x63F0BFACF2C00F6B
# > Sir
# #02  u64 [0]:	0xF76EA5B545A72761
# > Hashed
# #03  u64 [0]:	0x6CEC30CFBB0E4C7C
# > Sir
# #04  u64 [0]:	0xF76EA5B545A72761
# > Lol
# #05  u64 [0]:	0x68344BC6E6C587E0
# > Ctrl + D
# 1 Collisions
#
# Hashes #2 (Sir) and #4 (Sir) both get this hash:
# #01  u64 [0]:	0xF76EA5B545A72761
```

## Algorithms

Currently the following algorithms are supported:

- FNV-0: 32, 64, & 128 (`--fnv0-32`, `--fnv0-64`, `--fnv0-128`)
- FNV-1: 32, 64, & 128 (`--fnv1-32`, `--fnv1-64`, `--fnv1-128`)
- FNV-1A: 32, 64, & 128 (`--fnv1a-32`, `--fnv1a-64`, `--fnv1a-128`)

## Options

Here are all the options. You can also run `clhasher --help` for a similar list.

Usage:

```shell
clhasher <options> <algorithm option> <string>
```
Or omit `<string>` for STDIN:
```shell
echo 'hello' | clhasher <options> <algorithm option>
```

Miscellaneous:

- `-o <FILE>`/`--out=<FILE>` Print into a file. Defaults to `-` for STDOUT.
- `-i <FILE>`/`--in=<FILE>` Read from a file if no value argument is given. Defaults to `-` for STDIN.
- `-z <BYTES>`/`--length=<BYTES>` Number of bytes of input to hash. Defaults to `0` for automatic.
- `-0`/`--hash-nul` When length is automatic, includes the `\0` `NUL` to hash (by default does not hash the `NUL`).
- `-e`/`--escape` Allow escape sequences in the inputted string (see [Escaping](#escaping)).
- `-m`/`--multi` Individually hash each line of input (splits input by a line feed). Also automatically reports any collisions.
- `-v`/`--verbose` Describe what is happening.
- `-?`/`--help` Print help list.
- `--usage` See usage message.
- `-V`/`--version` Print program version.

[Algorithms](#algorithms):

- `--fnv0-32` FNV-0 32-bit algorithm.
- `--fnv0-64` FNV-0 64-bit algorithm.
- `--fnv0-128` FNV-0 128-bit algorithm.
- `--fnv1-32` FNV-1 32-bit algorithm.
- `--fnv1-64` FNV-1 64-bit algorithm.
- `--fnv1-128` FNV-1 128-bit algorithm.
- `--fnv1a-32` FNV-1A 32-bit algorithm.
- `--fnv1a-64` FNV-1A 64-bit algorithm.
- `--fnv1a-128` FNV-1A 128-bit algorithm.

Output Formatting:

- `-a`/`--annotate` Print output with nicer formatting.
- `--hi-to-lo` When splitting the output, print the most significant part first (i.e. simulates big-endian on an integer scale).
- `-x`/`--hex` Print base-16 hexadecimal numbers.
- `--octal` Print base-8 octal numbers.
- `--base-10` Print base-10 "normal" numbers (default).

Splitting:

- `--split-32` Split hash output into one/multiple 32-bit numbers (e.g. 64-bit prints two 32-bit numbers).
- `--split-64` Split hash output into one/multiple 64-bit numbers (default).

## Escaping

You can enable escaping via `-e`. Enabling escaping means that the string _must_
end at a `NUL`, and therefore having a string containing `NUL`s like
`-z <SIZE>` allows will not work.

| Escape Code    | ASCII | Description |
|----------------|:-----:|-------------|
| `\n`           | `0A`  | New line    |
| `\r`           | `0D`  | Carriage return |
| `\t`           | `09`  | Tab         |
| `\\`           | `5C`  | Backslash   |
| `\0`           | `00`  | Nul*        |
| `\XXX`         |       | Octal Code* |
| `\oXXX`        |       | Octal Code* |
| `\xXX`         |       | Hex Code*   |

*If a `NUL` character exists the string will end there.


CHANGELOG.md

## (Development) 1.1.0

- New `-m`/`--multi` mode with collision detection

## 1.0.0


src/appArgs.c
#include <argp.h>
#include <errno.h>
#include <stdio.h>
#include <stdbool.h>
#include <stdlib.h>
#include "algorithm.h"
#include "appArgs.h"

enum OptKey
{
    OPT_VERBOSE         = 'v',
    OPT_OUT             = 'o',
    OPT_FNV1A_32        = -1,
    OPT_FNV1A_64        = -10,
    OPT_FNV1A_128       = -11,
    OPT_FNV1_32         = -2,
    OPT_FNV1_64         = -12,
    OPT_FNV1_128        = -13,
    OPT_FNV0_32         = -3,
    OPT_FNV0_64         = -14,
    OPT_FNV0_128        = -15,
    OPT_b64             = -4,
    OPT_b32             = -5,
    OPT_b128            = -6,
    OPT_SPLIT_64        = 'q',
    OPT_SPLIT_32        = 'd',
    OPT_HI_TO_LO        = 'b',
    OPT_FLOAT           = 'l',
    OPT_DOUBLE          = 'f',
    OPT_HEX             = 'x',
    OPT_OCTAL           = -7,
    OPT_BINARY          = -8,
    OPT_BASE10          = -9,
    OPT_LEN             = 'z',
    OPT_HASH_NUL        = '0',
    OPT_ANNOTATE        = 'a',
    OPT_IN              = 'i',
    OPT_ESCAPE          = 'e',
    OPT_MULTI           = 'm'
};

const char *argp_program_version = "clhasher v1.0.0";
const char *argp_program_bug_address = "https://github.com/penguin-teal/clhasher";
static const char doc[] = "clhasher -- Quickly hash values.";
static const char argsDoc[] = "string";
static struct argp_option options[] =
{
    { 0         ,       0                   , 0     ,     0, "Output Options:", 1 },
    { "verbose" ,       OPT_VERBOSE         , 0     ,     0, "Describe what is happening.", 1 },
    { "out"     ,       OPT_OUT             , "FILE",     0, "Output to file. Defaults to '-' for STDOUT.", 1 },
    { "in"      ,       OPT_IN              , "FILE",     0, "Get input from file. Defaults to '-' for STDIN.", 1 },
    { "split-64",       OPT_SPLIT_64        , 0     ,     0, "Split the output to one 64-bit integer per line.", 1 },
    { "split-32",       OPT_SPLIT_32        , 0     ,     0, "Split the output to one 32-bit integer per line.", 1 },
    { "hi-to-lo",       OPT_HI_TO_LO        , 0     ,     0, "When splitting output, print the most significant part first.", 1 },
    //{ "float"   ,       OPT_FLOAT           , 0     ,     0, "Prints the output as (a) 32-bit IEEE 754 float(s) binary32.", 1 },
    //{ "double"  ,       OPT_DOUBLE          , 0     ,     0, "Prints the output as (a) 64-bit IEEE 754 double float(s) binary64.", 1 },
    { "length"  ,       OPT_LEN             , "SIZE",     0, "The size in bytes of the data to hash. Defaults to 0 which stops at NUL.", 1 },
    { "hash-nul",       OPT_HASH_NUL        , 0     ,     0, "When using auto length, includes the terminating NUL (\\0) in the hash.", 1 },
    { "annotate",       OPT_ANNOTATE        , 0     ,     0, "Annotate output with radix prefixes and index/size info.", 1 },
    { "escape",         OPT_ESCAPE          , 0     ,     0, "Character escape sequences.", 1 },
    { "multi",          OPT_MULTI           , 0     ,     0, "Split into multiple strings to hash by line feed ('\n', ASCII 0xA).", 1 },

    { 0         ,       0                   , 0     ,     0, "Print Number Format:", 2 },
    { "hex"     ,       OPT_HEX             , 0     ,     0, "Prints as (a) hexadecimal base-16 number(s).", 2 },
    { "octal"   ,       OPT_OCTAL           , 0     ,     0, "Prints as (a) octal base-8 number(s).", 2 },
    //{ "binary"  ,       OPT_BINARY          , 0     ,     0, "Prints as (a) binary base-2 number(s).", 2 },
    { "base-10" ,       OPT_BASE10          , 0     ,     0, "Default. Prints as (a) normal base-10 number(s).", 2 },
    
    { 0          ,       0                   , 0     ,     0, "Algorithms:", 4 },
    { "fnv1a-32" ,       OPT_FNV1A_32        , 0     ,     0, "Default. Use FNV-1a 32-bit hashing algorithm.", 4 },
    { "fnv1a-64" ,       OPT_FNV1A_64        , 0     ,     0, "Default. Use FNV-1a 64-bit hashing algorithm.", 4 },
    { "fnv1a-128",       OPT_FNV1A_128       , 0     ,     0, "Default. Use FNV-1a 128-bit hashing algorithm.", 4 },
    { "fnv1-32"  ,       OPT_FNV1_32         , 0     ,     0, "Use FNV-1 32-bit hashing algorithm.", 4 },
    { "fnv1-64"  ,       OPT_FNV1_64         , 0     ,     0, "Use FNV-1 64-bit hashing algorithm.", 4 },
    { "fnv1-128" ,       OPT_FNV1_128        , 0     ,     0, "Use FNV-1 128-bit hashing algorithm.", 4 },
    { "fnv0-32"  ,       OPT_FNV0_32         , 0     ,     0, "Use FNV-0 32-bit hashing algorithm.", 4 },
    { "fnv0-64"  ,       OPT_FNV0_64         , 0     ,     0, "Use FNV-0 64-bit hashing algorithm.", 4 },
    { "fnv0-128" ,       OPT_FNV0_128        , 0     ,     0, "Use FNV-0 128-bit hashing algorithm.", 4 },

    { 0         ,       0                   , 0     ,     0, "Miscellaneous:", -1 },

    { 0 }
};

static error_t parseOpt(int key, char *arg, struct argp_state *state)
{
    struct AppArgs *arguments = state->input;

    switch(key)
    {
        case OPT_VERBOSE:
            arguments->verbose = true;
            break;
        case OPT_OUT:
            arguments->out = arg;
            break;
        case OPT_IN:
            arguments->in = arg;
            break;
        case OPT_SPLIT_64:
            arguments->splitBits = 64;
            break;
        case OPT_SPLIT_32:
            arguments->splitBits = 32;
            break;
        case OPT_HI_TO_LO:
            arguments->hiToLo = true;
            break;
        case OPT_LEN:
            {
                int oldErrno = errno;
                errno = 0;

                char *end;
                uint64_t len = strtoul(arg, &end, 10);
                if(!len && errno)
                {
                    errno = oldErrno;
                    argp_usage(state);
                    break;
                }
                else
                {
                    errno = oldErrno;
                    arguments->len = len;
                    break;
                }
            }
        case OPT_HASH_NUL:
            arguments->hashNul = true;
            break;
        case OPT_ANNOTATE:
            arguments->annotate = true;
            break;
        case OPT_ESCAPE:
            arguments->escape = true;
            break;
        case OPT_MULTI:
            arguments->multi = true;
            break;

        case OPT_BINARY:
            arguments->radix = 2;
            break;
        case OPT_OCTAL:
            arguments->radix = 8;
            break;
        case OPT_HEX:
            arguments->radix = 16;
            break;
        case OPT_BASE10:
            arguments->radix = 10;
            break;

        case OPT_FNV1A_32:
            arguments->algorithm = ALG_FNV1A_32;
            break;
        case OPT_FNV1A_64:
            arguments->algorithm = ALG_FNV1A_64;
            break;
        case OPT_FNV1A_128:
            arguments->algorithm = ALG_FNV1A_128;
            break;
        case OPT_FNV1_32:
            arguments->algorithm = ALG_FNV1_32;
            break;
        case OPT_FNV1_64:
            arguments->algorithm = ALG_FNV1_64;
            break;
        case OPT_FNV1_128:
            arguments->algorithm = ALG_FNV1_128;
            break;
        case OPT_FNV0_32:
            arguments->algorithm = ALG_FNV0_32;
            break;
        case OPT_FNV0_64:
            arguments->algorithm = ALG_FNV0_64;
            break;
        case OPT_FNV0_128:
            arguments->algorithm = ALG_FNV0_128;
            break;
        
        case OPT_b64:
            arguments->bits = 64;
            break;
        case OPT_b32:
            arguments->bits = 32;
            break;
        case OPT_b128:
            arguments->bits = 128;
            break;
        
        case ARGP_KEY_ARG:
            if(state->arg_num > 1)
            {
                argp_usage(state);
            }
            arguments->value = arg;
            break;
        case ARGP_KEY_END:
            break;

        default:
            return ARGP_ERR_UNKNOWN;
    }

    return 0;
}

static struct argp argp =
{
    .options = options,
    .parser = parseOpt,
    .args_doc = argsDoc,
    .doc = doc
};

bool doArgp(struct AppArgs *appArgs, int argc, char **argv)
{
    appArgs->value = NULL;
    appArgs->verbose = false;
    appArgs->out = "-";
    appArgs->in = "-";
    appArgs->algorithm = 0;
    appArgs->bits = 64;
    appArgs->splitBits = 64;
    appArgs->hiToLo = false;
    appArgs->radix = 10;
    appArgs->len = 0;
    appArgs->hashNul = false;
    appArgs->annotate = false;
    appArgs->escape = false;
    appArgs->multi = false;

    error_t result = argp_parse(&argp, argc, argv, 0, 0, appArgs);

    appArgs->bits = appArgs->algorithm & ALG_SIZE_MASK;

    if(!appArgs->splitBits || appArgs->splitBits > appArgs->bits)
    {
        appArgs->splitBits = appArgs->bits;
    }
    
    return !result;
}
